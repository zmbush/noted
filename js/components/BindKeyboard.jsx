// Copyright 2019 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

import React from 'react';
import Mousetrap from 'mousetrap';
import PropTypes from 'prop-types';

export default class BindKeyboard extends React.Component {
  static propTypes = {
    keys: PropTypes.string.isRequired,
    action: PropTypes.string,
    callback: PropTypes.func.isRequired,
    global: PropTypes.bool,
  };

  constructor(props) {
    super(props);

    this.main = React.createRef();
  }

  componentDidMount() {
    if (this.props.children) {
      this.mousetrap = new Mousetrap(this.main.current);
    } else {
      this.mousetrap = Mousetrap;
    }

    this.mousetrap.bind(
      this.props.keys,
      this.props.callback,
      this.props.action
    );
  }

  componentWillUnmount() {
    this.mousetrap.unbind(this.props.keys, this.props.action);
  }

  render() {
    if (this.props.children) {
      return <div ref={this.main}>{this.props.children}</div>;
    } else {
      return null;
    }
  }
}
