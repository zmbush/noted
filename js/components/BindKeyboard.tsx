// Copyright 2019 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.
import Mousetrap from 'mousetrap';

import * as React from 'react';

type Props = {
  keys: string;
  action?: string;
  callback: (e: Mousetrap.ExtendedKeyboardEvent, combo: string) => void;
};

export default class BindKeyboard extends React.Component<Props> {
  main: React.RefObject<HTMLDivElement>;

  mousetrap: Mousetrap.MousetrapInstance | Mousetrap.MousetrapStatic;

  constructor(props: Props) {
    super(props);

    this.main = React.createRef();
  }

  componentDidMount() {
    const { children, keys, callback, action } = this.props;
    if (children) {
      this.mousetrap = new Mousetrap(this.main.current);
    } else {
      this.mousetrap = Mousetrap;
    }

    this.mousetrap.bind(keys, callback, action);
  }

  componentWillUnmount() {
    const { keys, action } = this.props;
    this.mousetrap.unbind(keys, action);
  }

  render() {
    const { children } = this.props;
    if (children) {
      return <div ref={this.main}>{children}</div>;
    }
    return null;
  }
}
