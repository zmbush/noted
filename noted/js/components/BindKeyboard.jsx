// Copyright 2019 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

import React from 'react';
import Mousetrap from 'mousetrap';
import PropTypes from 'prop-types';

let _globalCallbacks = {};
let _originalStopCallback = Mousetrap.prototype.stopCallback;

Object.assign(Mousetrap.prototype, {
  stopCallback(e, element, keys, sequence) {
    if (this.paused) {
      return true;
    }

    if (_globalCallbacks[keys] || _globalCallbacks[sequence]) {
      return false;
    }

    return _originalStopCallback.call(this, e, element, keys);
  },

  bindGlobal(keys, callback, action) {
    this.bind(keys, callback, action);

    if (keys instanceof Array) {
      for (var i = 0; i < keys.length; i++) {
        _globalCallbacks[keys[i]] = true;
      }
      return;
    }

    _globalCallbacks[keys] = true;
  },

  unbindGlobal(keys, action) {
    this.unbind(keys, action);

    if (keys instanceof Array) {
      for (var i = 0; i < keys.length; i++) {
        delete _globalCallbacks[keys[i]];
      }
      return;
    }

    delete _globalCallbacks[keys];
  },
});

Mousetrap.init();

export default class BindKeyboard extends React.Component {
  static propTypes = {
    keys: PropTypes.string.isRequired,
    action: PropTypes.string,
    callback: PropTypes.func.isRequired,
    global: PropTypes.bool,
  };

  constructor(props) {
    super(props);

    this.main = React.createRef();
  }

  componentDidMount() {
    if (this.props.children) {
      this.mousetrap = new Mousetrap(this.main.current);
    } else {
      this.mousetrap = Mousetrap;
    }

    this.mousetrap.bind(
      this.props.keys,
      this.props.callback,
      this.props.action
    );
  }

  componentWillUnmount() {
    this.mousetrap.unbind(this.props.keys, this.props.action);
  }

  render() {
    if (this.props.children) {
      return <div ref={this.main}>{this.props.children}</div>;
    } else {
      return null;
    }
  }
}
