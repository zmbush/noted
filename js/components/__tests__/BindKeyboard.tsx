// Copyright 2019 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

import * as React from 'react';
import { shallow } from 'enzyme';
import Mousetrap from 'mousetrap';
//jest.mock('mousetrap');

const MockedMousetrap = (Mousetrap as any) as jest.Mock<typeof Mousetrap>;

import BindKeyboard from '../BindKeyboard';

describe('<BindKeyboard />', () => {
  test('matches snapshot', () => {
    expect(
      shallow(<BindKeyboard keys={'key'} callback={() => {}} />)
    ).toMatchSnapshot();

    expect(
      shallow(
        <BindKeyboard keys={'key'} callback={() => {}}>
          Contents
        </BindKeyboard>
      )
    ).toMatchSnapshot();
  });

  test('unmount works', () => {
    const bindFn = MockedMousetrap.mock.instances[0].bind;
    const mockedBind = bindFn as jest.Mock<typeof bindFn>;
    mockedBind.mockClear();

    const unbindFn = MockedMousetrap.mock.instances[0].unbind;
    const mockedUnbind = unbindFn as jest.Mock<typeof unbindFn>;
    mockedUnbind.mockClear();

    let cb = () => {};

    const wrapper = shallow(
      <BindKeyboard keys='a+key' callback={cb} action='toot' />
    );
    expect(MockedMousetrap.mock.instances[0].bind).toBeCalledWith(
      'a+key',
      cb,
      'toot'
    );
    wrapper.unmount();
    expect(MockedMousetrap.mock.instances[0].unbind).toBeCalledWith(
      'a+key',
      'toot'
    );
  });
});
