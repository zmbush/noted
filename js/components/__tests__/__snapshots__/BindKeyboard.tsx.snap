// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BindKeyboard /> matches snapshot 1`] = `""`;

exports[`<BindKeyboard /> matches snapshot 2`] = `
<div>
  Contents
</div>
`;
